core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java

```diff
@@ -65,19 +69,44 @@ public PreparedQuery prepareQuery(Session session, Statement wrappedStatement)
             prepareSql = Optional.of(session.getPreparedStatementFromExecute(executeStatement));
             statement = sqlParser.createStatement(prepareSql.get(), createParsingOptions(session));
         }
-
-        if (statement instanceof ExplainAnalyze explainAnalyzeStatement) {
+        else if (statement instanceof ExecuteImmediate executeImmediateStatement) {
+            try {
+                statement = sqlParser.createStatement(
+                        executeImmediateStatement.getStatement().getValue(),
+                        executeImmediateStatement.getStatement().getLocation().orElseThrow(() -> new ParsingException("Missing location for embedded statement")),
+                        createParsingOptions(session));
+            }
+            catch (ParsingException e) {
+                NodeLocation relativeLocation = executeImmediateStatement.getStatement().getLocation().orElse(new NodeLocation(1, 1));
+                int embeddedStatementLine = relativeLocation.getLineNumber();
+                int embeddedStatementColumn = relativeLocation.getColumnNumber();
+                throw new TrinoException(
+                        SYNTAX_ERROR,
+                        new ParsingException(
+                                e.getErrorMessage(),
+                                (RecognitionException) e.getCause(),
+                                e.getLineNumber() + embeddedStatementLine - 1,
+                                e.getColumnNumber() + (embeddedStatementLine == 1 ? embeddedStatementColumn : 0))
+                                .getMessage());
+            }
+        }
```

Comment:
This should not be needed anymore. It should be taken care of by the use of createStatement with a base location above.

Reply:
This is needed to handle syntax errors.  This is why we would ideally make the lexer aware of the relative parsing position.

Reply:
That should be handled in the SqlParser class, though.

Reply:
Oh, good point.